django/db/models/expressions.py

```diff
@@ -287,7 +287,8 @@ def _resolve_output_field(self):
         sources_iter = (source for source in self.get_source_fields() if source is not None)
         for output_field in sources_iter:
             if any(not isinstance(output_field, source.__class__) for source in sources_iter):
-                raise FieldError('Expression contains mixed types. You must set output_field.')
+                raise FieldError('Expression contains mixed types: %s, %s. You must set output_field to %s.'
+                        %(output_field.__class__.__name__, source.__class__.__name__, source.__class__.__name__)
```

Comment:
Use hanging indentation
```python
raise FieldError(
    'Expression contains mixed types: %s, %s. You must set '
    'output_field to %s.' % (
        output_field.__class__.__name__,
        source.__class__.__name__,
        source.__class__.__name__,
    )
)
```

Reply:
Ignoring the SyntaxError and overall formatting, `source` isn't in scope (it's in a genexp), only `output_field` is, unless you decompose the genexp into local variables.

But if the solution implies that it's always a clear-cut "you must use this field" (rather than one of N possible output field handlers) ... why does this exception even exist?

Reply:
Just realized. don't know why my local environment doesn't show any errors. 

Maybe we could just output the fields in the expression and not suggest an output field. At least they'll be a quick realization of the wrong fields inputted. Someone asked for it.

Reply:
@kezabelle Agreed, we shouldn't suggest an `output_type`, so the second sentence should not be changed.

Reply:
Maybe:
```diff
diff --git a/django/db/models/expressions.py b/django/db/models/expressions.py                                                                         
index 16df317631..36f88f99ec 100644                                                                                                                    
--- a/django/db/models/expressions.py                                                                                                                  
+++ b/django/db/models/expressions.py                                                                                                                  
@@ -286,8 +286,15 @@ class BaseExpression:                                                                                                             
         """                                                                                                                                           
         sources_iter = (source for source in self.get_source_fields() if source is not None)                                                          
         for output_field in sources_iter:                                                                                                             
-            if any(not isinstance(output_field, source.__class__) for source in sources_iter):                                                        
-                raise FieldError('Expression contains mixed types. You must set output_field.')                                                       
+            for source in sources_iter:                                                                                                               
+                if not isinstance(output_field, source.__class__):                                                                                    
+                    raise FieldError(                                                                                                                 
+                        'Expression contains mixed types: %s, %s. You must '                                                                          
+                        'set output_field.' % (                                                                                                       
+                            output_field.__class__.__name__,                                                                                          
+                            source.__class__.__name__,                                                                                                
+                        )                                                                                                                             
+                    )                                                                                                                                 
             return output_field                                                                                                                       
                                                                                                                                                       
     @staticmethod
```

Reply:
@CruxBox Please fix also `aggregation.tests.AggregateTestCase.test_combine_different_types`.